{"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"name":"python","version":"3.7.8","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"# COMP90014 Assignment 1\n### Semester 2, 2020","metadata":{}},{"cell_type":"markdown","source":"This assignment should be completed by each student individually. Make sure you read this entire document, and ask for help if anything is not clear. Any changes or clarification to this document will be announced via Canvas.\n\nPlease make sure you are aware of the University's rules on academic honesty and plagiarism, which are very strict: https://academichonesty.unimelb.edu.au/ \n\nMake sure you **do not** copy any code either from other students or from the internet. This is considered plagiarism. It is generally a good idea to avoid looking at any solutions as you may find it surprisingly difficult to generate your own solution to the problem once you have seen somebody else's.\n\nYour completed notebook file containing all your answers will be turned in via Canvas. No other files or formats will be accepted - only upload the completed `.ipynb` file.","metadata":{}},{"cell_type":"markdown","source":"### Overview\nTo complete the assignment you will need to finish the tasks in this notebook. There are multiple tasks that are connected in a logical order.\n\nThe tasks are a combination of writing your own implementations of algorithms we've discussed in lectures, writing your own code to use library implementations of these algorithms and interpreting the results in short answer format. Each short answer question has a word limit that will be strictly enforced!\n\nIn some case, we have provided test input and test output that you can use to try out your solutions. These tests are just samples and are not exhaustive - they may warn you if you've made a mistake, but they are not guaranteed to. It's up to you to decide whether your code is correct.","metadata":{}},{"cell_type":"markdown","source":"### Marking\n\nCells that must be completed to receive marks are clearly labeled. There are 16 graded cells, some of which are code cells, in which you must complete the code to solve a problem, and some of which are markdown cells, in which you must write your answers to short-answer questions. In order to get full marks, your code needs to be clear, unambiguous, with adequate comments. **Code style, readability and efficiency** will be taken into consideration. \n\nThe total marks for the assignment add up to 45, and it will be worth 15% of your overall subject grade.","metadata":{}},{"cell_type":"markdown","source":"### Background and data\nDuring this assignment we will analyse the relationship between 17 organisms. \nUnderstanding the genetic and evolutionary relationship between organisms is an important task which is at the core of bioinformatics. \nThis type of phylogenetic analysis can be performed by comparing the sequence differences of a particular gene between the organisms.\nAll organisms must possess the gene in question for comparison to be possible.  \n\nAs all organisms possess genes to make ribosomes, these will be used in this assignment.\nA 100bp segment of the small ribosomal subunit gene has been extracted from each of these 17 organisms, and will act as our sequences for comparison.\n\nIn the first section, we will calculate distances between these organisms and visualise.\nIn the second section, we will use clustering algorithms to group organisms.\n\nAt each step of the process you will be asked to fill out short answer questions to demonstrate your understanding of the concepts.","metadata":{}},{"cell_type":"markdown","source":"### Package installs\nIf you are using jupyter lab online, all packages will be available. If you are running this on your local computer, you may need to install some packages. Run the cell below if using jupyter lab locally. ","metadata":{}},{"cell_type":"code","source":"!pip install numpy\n!pip install scipy\n!pip install pandas\n!pip install scikit-learn\n!pip install seaborn","metadata":{"collapsed":true,"jupyter":{"outputs_hidden":true}},"execution_count":3,"outputs":[{"name":"stdout","output_type":"stream","text":"Requirement already satisfied: numpy in /srv/conda/envs/notebook/lib/python3.7/site-packages (1.19.1)\nRequirement already satisfied: scipy in /srv/conda/envs/notebook/lib/python3.7/site-packages (1.5.2)\nRequirement already satisfied: numpy>=1.14.5 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from scipy) (1.19.1)\nRequirement already satisfied: pandas in /srv/conda/envs/notebook/lib/python3.7/site-packages (0.24.2)\nRequirement already satisfied: python-dateutil>=2.5.0 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from pandas) (2.8.1)\nRequirement already satisfied: numpy>=1.12.0 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from pandas) (1.19.1)\nRequirement already satisfied: pytz>=2011k in /srv/conda/envs/notebook/lib/python3.7/site-packages (from pandas) (2020.1)\nRequirement already satisfied: six>=1.5 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from python-dateutil>=2.5.0->pandas) (1.15.0)\nRequirement already satisfied: scikit-learn in /srv/conda/envs/notebook/lib/python3.7/site-packages (0.21.3)\nRequirement already satisfied: numpy>=1.11.0 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from scikit-learn) (1.19.1)\nRequirement already satisfied: scipy>=0.17.0 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from scikit-learn) (1.5.2)\nRequirement already satisfied: joblib>=0.11 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from scikit-learn) (0.16.0)\nRequirement already satisfied: seaborn in /srv/conda/envs/notebook/lib/python3.7/site-packages (0.9.0)\nRequirement already satisfied: matplotlib>=1.4.3 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from seaborn) (3.1.3)\nRequirement already satisfied: pandas>=0.15.2 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from seaborn) (0.24.2)\nRequirement already satisfied: numpy>=1.9.3 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from seaborn) (1.19.1)\nRequirement already satisfied: scipy>=0.14.0 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from seaborn) (1.5.2)\nRequirement already satisfied: python-dateutil>=2.1 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from matplotlib>=1.4.3->seaborn) (2.8.1)\nRequirement already satisfied: cycler>=0.10 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from matplotlib>=1.4.3->seaborn) (0.10.0)\nRequirement already satisfied: kiwisolver>=1.0.1 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from matplotlib>=1.4.3->seaborn) (1.2.0)\nRequirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from matplotlib>=1.4.3->seaborn) (2.4.7)\nRequirement already satisfied: pytz>=2011k in /srv/conda/envs/notebook/lib/python3.7/site-packages (from pandas>=0.15.2->seaborn) (2020.1)\nRequirement already satisfied: six>=1.5 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from python-dateutil>=2.1->matplotlib>=1.4.3->seaborn) (1.15.0)\n"}]},{"cell_type":"markdown","source":"### Importing data\nTo begin, we need to understand the data.\nThe ribosome genes are available in a .fasta file called 'ribosome_genes.fasta'. You can have a look if you like.\nThese genes will be imported as classes (RibosomeGene). \n\nEach RibosomeGene object has a name, accession, sequence and length.\nYou can access these properties using '.' syntax. (see below). \n\nTry to think of each gene as something physical, rather than code.\nIn real life, each gene has a .length, its organism has a .name, and it has a .sequence. We can write code in this way too. \n\nwe will import these into **ribosome_genes**, which is a list of our genes.","metadata":{}},{"cell_type":"code","source":"import warnings\nwarnings.filterwarnings('ignore')\n\nfrom utilities import import_16s_sequences\nribosome_genes = import_16s_sequences()","metadata":{},"execution_count":1,"outputs":[]},{"cell_type":"code","source":"print('{:<20}{:<30}{:<15}{:<10}'.format('gene.accession', 'gene.name', 'gene.length', 'gene.sequence'))\nfor gene in ribosome_genes:\n    print('{:<20}{:<30}{:<15}{:<10}'.format(gene.accession, gene.name[:27], gene.length, gene.sequence[:8] + '...'))","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### SECTION 1: PAIRWISE DISTANCES\nTo be able to compare organisms via their sequences, we need a way to measure their difference as a distance. ","metadata":{}},{"cell_type":"markdown","source":"**TASK 1: K-mer distance**<br>\nThe k-mer distance between two sequences is the fraction of k-mers that are unique to either sequence\nWrite a function that calculates a dictionary of k-mers (for k = any number) and their counts for each gene sequence.","metadata":{}},{"cell_type":"code","source":"# -- GRADED CELL (1 mark) - complete this cell --\n\ndef create_kmer_dictionary(seq, k):\n    # write a function which stores kmers, and their counts from a given input seq. \n    # the kmer size == k\n    \n    return kmer_dict","metadata":{},"execution_count":3,"outputs":[]},{"cell_type":"markdown","source":"Write a function that accepts two k-mer occurance dictionaries, and returns the k-mer distance","metadata":{}},{"cell_type":"code","source":"# -- GRADED CELL (2 marks) - complete this cell --\n\ndef calculate_total_unique_kmers(kmers1, kmers2):\n    # given two kmer dictionaries, compute the sum of kmers unique to either dictionary\n\n    return unique_kmers","metadata":{},"execution_count":4,"outputs":[]},{"cell_type":"markdown","source":"Write a function that uses the above two functions to calcualte the k-mer distance of two sequences","metadata":{}},{"cell_type":"code","source":"# -- GRADED CELL (2 marks) - complete this cell --\n\ndef kmer_distance(seq1, seq2, k):\n    # for two input sequences, find their kmer distance.\n    # use the create_kmer_dictionary and calculate_total_unique_kmers functions you have written previously. \n    \n    return distance","metadata":{},"execution_count":5,"outputs":[]},{"cell_type":"markdown","source":"Let's check our function. We can use the first two entries in the 'ribosome_genes' list.\nIf implemented correctly, the following should return 24","metadata":{}},{"cell_type":"code","source":"distance = kmer_distance(ribosome_genes[1].sequence, ribosome_genes[3].sequence, 8)\nprint(distance)","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"**TASK2: smith-waterman alignment**<br>\nAnother way to compare the similarity of two sequences is through alignment.\nThe alignment score of two sequences will be high when they are similar, and low when they are distinct. ","metadata":{}},{"cell_type":"markdown","source":"Write a function that initialises and returns a scoregrid for two sequences. <br>\nKeep in mind the matrix must be 1 element larger than the sequence lengths. Consider whether indel scores for the first row and column need to be filled in.","metadata":{}},{"cell_type":"code","source":"# -- GRADED CELL (1 mark) - complete this cell --\n\nimport numpy as np\ndef init_scoregrid(seq1, seq2, indel_score=-4):\n    # given two sequences, initialise a scoregrid to be used for smith-waterman alignment.\n    \n    return scoregrid","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Let's do a sanity check that the grid has been initialised properly. <br>\nThe following should print the initialised scoregrid","metadata":{}},{"cell_type":"code","source":"print(init_scoregrid('hello', 'kittycat'))","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Write a function that calculates the initialised scoregrid. It accepts two sequences, a scoregrid and match/mismatch and indel scores. ","metadata":{}},{"cell_type":"code","source":"# -- GRADED CELL (2 marks) - complete this cell --\n\ndef calculate_scoregrid(seq1, seq2, scoregrid, match_score=1, mismatch_score=-4, indel_score=-4):\n    # given two sequences and an initialised scoregrid, fill in the cumulative alignment scores for each cell.\n\n    return scoregrid","metadata":{},"execution_count":79,"outputs":[]},{"cell_type":"markdown","source":"Let's do another sanity check. <br>\nThe following should print a calculated scoregrid, with the these numbers in the bottom right corner: <br>\n 0  0 <br>\n-4  1","metadata":{}},{"cell_type":"code","source":"scoregrid = init_scoregrid('hello', 'helllo')\nprint(calculate_scoregrid('hello', 'helllo', scoregrid))","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"The last thing we need is to get the alignment score from the scoregrid. \nWrite a function that returns an alignment score (smith-waterman style, local alignment) from a scoregrid","metadata":{}},{"cell_type":"code","source":"# -- GRADED CELL (1 marks) - complete this cell --\n\ndef report_alignment_score(scoregrid):\n    # given a completed scoregrid, return the smith-waterman alignment score.\n    \n    return sw_alignment_score","metadata":{},"execution_count":81,"outputs":[]},{"cell_type":"markdown","source":"Final sanity check. Should return 4.","metadata":{}},{"cell_type":"code","source":"scoregrid = init_scoregrid('hello', 'helllo')\ncalculated_scoregrid = calculate_scoregrid('hello', 'helllo', scoregrid)\nprint(report_alignment_score(calculated_scoregrid))","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Ok! now we're ready to put it all together. <br>\nFill in the function below with the three functions you wrote to calculate the alignment score of two sequences","metadata":{}},{"cell_type":"code","source":"# -- GRADED CELL (1 marks) - complete this cell --\n\ndef smith_waterman(seq1, seq2):\n    # Given two sequences, perform smith-waterman alignment score calculation. \n    # use the init_scoregrid, calculate_scoregrid, and report_alignment_score functions you have written.\n    \n    return alignment_score","metadata":{},"execution_count":83,"outputs":[]},{"cell_type":"markdown","source":"The following should print 4","metadata":{}},{"cell_type":"code","source":"print(smith_waterman('hello', 'helllo'))","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"**TASK 3: pairwise distances**","metadata":{}},{"cell_type":"markdown","source":"We have now written two functions which can calculate the distance of two sequences. \nWe can calculate the k-mer distance, and the smith-waterman alignment score. \nlets use these two methods to calculate the pairwise distance of our genes. ","metadata":{}},{"cell_type":"markdown","source":"Write a function which initialises and returns a distance matrix from a list of genes. \nUse a nested dictionary, where keys are 'gene.accession's, and the values are all 0 for now. ","metadata":{}},{"cell_type":"code","source":"# -- GRADED CELL (1 mark) - complete this cell --\n\ndef init_distance_matrix(genes):\n    # for our list of genes, initialise a pairwise-distance matrix. Use accessions to denote genes. \n    \n    return distance_matrix","metadata":{},"execution_count":85,"outputs":[]},{"cell_type":"markdown","source":"Let's print the distance matrix to make sure it worked. ","metadata":{}},{"cell_type":"code","source":"from utilities import print_distance_matrix\ndistance_matrix = init_distance_matrix(ribosome_genes)\nprint_distance_matrix(distance_matrix)","metadata":{},"execution_count":86,"outputs":[{"name":"stdout","output_type":"stream","text":"        EC_A  EC_B  EC_C  EC_D  EC_E  EC_F  EC_G  EF_A  EF_B  EF_C  EF_D  HS_A  MI_A  SD_A  SF_A  SS_A  SS_B\nEC_A     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nEC_B     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nEC_C     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nEC_D     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nEC_E     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nEC_F     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nEC_G     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nEF_A     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nEF_B     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nEF_C     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nEF_D     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nHS_A     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nMI_A     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nSD_A     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nSF_A     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nSS_A     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\nSS_B     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\n"}]},{"cell_type":"markdown","source":"Time to fill in the matrix with distances. <br>\nWrite a function which calculates the pairwise distance of genes using kmer distance.\nyou will need to call the 'kmer_distance' function you have written above. ","metadata":{}},{"cell_type":"code","source":"# -- GRADED CELL (2 marks) - complete this cell --\n\ndef calculate_kmer_distance_matrix(genes, matrix, k):\n    # using the initialised distance matrix, our list of genes, and your kmer_distance function,\n    # populate the matrix with kmer distances between each pair of genes. \n    \n    return matrix","metadata":{},"execution_count":87,"outputs":[]},{"cell_type":"markdown","source":"Let's do the same as above, but this time use the 'smith_waterman' alignment distance function you wrote. ","metadata":{}},{"cell_type":"code","source":"# -- GRADED CELL (1 marks) - complete this cell --\n\ndef calculate_sw_alignment_distance_matrix(genes, matrix):\n    # using the initialised distance matrix, our list of genes, and your smith_waterman function,\n    # populate the matrix with smith-waterman alignment scores between each pair of genes. \n    \n    return matrix","metadata":{},"execution_count":88,"outputs":[]},{"cell_type":"markdown","source":"Let's test them out. The two cells below will use your calculate_kmer_distance_matrix, and calculate_sw_alignment_distance_matrix functions to add distances to the matrix. <br>\n**NOTE:** the smith-waterman distance calculations can take time. Give it a minute.","metadata":{}},{"cell_type":"code","source":"distance_matrix = init_distance_matrix(ribosome_genes)\nkmer_distance_matrix = calculate_kmer_distance_matrix(ribosome_genes, distance_matrix, 8)\nprint('\\nkmer distance matrix')\nprint_distance_matrix(kmer_distance_matrix)","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"distance_matrix = init_distance_matrix(ribosome_genes)\nsw_alignment_distance_matrix = calculate_sw_alignment_distance_matrix(ribosome_genes, distance_matrix)\nprint('\\nsmith waterman alignment score distance matrix')\nprint_distance_matrix(sw_alignment_distance_matrix)","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Let's visualise those in a better manner for human eyes. The cell below will plot heatmaps instead of raw numbers.","metadata":{}},{"cell_type":"code","source":"from utilities import heatmap\nheatmap(kmer_distance_matrix, sw_alignment_distance_matrix)","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Comment in the cell below why these two heatmaps look inverted relative to one another. <br>\nComment on the extent of similarity between the Smith-Waterman and k-mer distance approaches, and suggest which may be more appropriate for biological sequences, making references to some knowledge you have about genes. <br> \n(Maximum of 150 words)","metadata":{}},{"cell_type":"markdown","source":" -- GRADED CELL (6 marks) - complete this cell --\n ","metadata":{}},{"cell_type":"markdown","source":"### SECTION 2: CLUSTERING","metadata":{}},{"cell_type":"markdown","source":"From the heatmaps, it seems like there are a few clusters in the data. <br>\nFirst, lets convert the pairwise distances to 2D coordinates. \nThis is possible using Multidimensional scaling (MDS - which we will cover in a few weeks during lectures).\nAfter we have transformed the distance matrix to 2D coordinates, we can plot it to see if any clusters are evident.","metadata":{}},{"cell_type":"code","source":"from utilities import mds_scatterplot, distance_matrix_to_coordinates_MDS\nkmer_distances_xy = distance_matrix_to_coordinates_MDS(kmer_distance_matrix)\nsw_distances_xy = distance_matrix_to_coordinates_MDS(sw_alignment_distance_matrix)\nmds_scatterplot(kmer_distances_xy)\nmds_scatterplot(sw_distances_xy)","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Seems like there is some clustering happening. <br>\nLet's use some clustering algorithms to define the clusters. \nin this manner, we can have an objective way to talk about the patterns in the data.","metadata":{}},{"cell_type":"markdown","source":"**TASK 4: K-means & K-medoids**\nwe are going to use K-means and K-medoids to cluster the data.","metadata":{}},{"cell_type":"markdown","source":"Let's implement the k-means algorithm. <br>\nWe have provided initialise_centroids, assign_points, and k-means. \nwrite a function which calculates new centroid locations (using the mean)","metadata":{}},{"cell_type":"code","source":"from utilities import initialise_centroids, average_point, assign_points, plot_kmeans, points_equal, euclidean_distance","metadata":{},"execution_count":115,"outputs":[]},{"cell_type":"code","source":"# -- GRADED CELL (1 marks) - complete this cell --\n\ndef calculate_mean_centroids(data, assignments, k):\n    # Given a list of points (data), and their corresponding cluster assignments (assignments), \n    # update our centroid locations by calculating the mean location of points for each cluster. \n        \n    return centroids","metadata":{},"execution_count":116,"outputs":[]},{"cell_type":"markdown","source":"Place calculate_mean_centroids() in the kmeans function below to complete kmeans","metadata":{}},{"cell_type":"code","source":"# -- GRADED CELL (2 marks) - complete this cell --\n\ndef kmeans(data, k):\n    # given a list of points (data) and num clusters = k, perform k-means clustering\n    \n        \n    #optional\n    #plot_kmeans(data, centroids, cluster_assignments, k)\n    \n    return centroids, cluster_assignments","metadata":{},"execution_count":117,"outputs":[]},{"cell_type":"markdown","source":"You can check your implementation using the cell below:","metadata":{}},{"cell_type":"code","source":"centroids, cluster_assignments = kmeans(kmer_distances_xy, 3)\nplot_kmeans(kmer_distances_xy, centroids, cluster_assignments, 3)","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Let's also implement k-medoids while we're at it. <br>\nThe only difference between k-means and k-medoids is the calculate_mean_centroids() step, which will instead be calculate_median_centroids()\nthe median can be taken here as the point in the cluster which has smallest cumulative distance to the other points in the cluster\nYou can use the provided euclidean_distance() function to calculate distances between points\nwrite a function which calculates new centroid locations (using the median)","metadata":{}},{"cell_type":"code","source":"# -- GRADED CELL (3 marks) - complete this cell --\n\ndef calculate_median_centroids(data, assignments, k):\n    # Given a list of points (data), and their corresponding cluster assignments (assignments), \n    # update our centroid locations by calculating the median point of each cluster.  \n    \n    return centroids","metadata":{},"execution_count":119,"outputs":[]},{"cell_type":"markdown","source":"Place calculate_median_centroids() in the kmedoids function below to complete kmedoids","metadata":{}},{"cell_type":"code","source":"# -- GRADED CELL (1 marks) - complete this cell --\n\ndef kmedoids(data, k):\n    # given a list of points (data) and num clusters = k, perform k-means clustering\n    \n\n    #optional\n    #plot_kmeans(data, centroids, cluster_assignments, k)\n    \n    return centroids, cluster_assignments","metadata":{},"execution_count":120,"outputs":[]},{"cell_type":"markdown","source":"Here is another check cell, for kmedoids this time:","metadata":{}},{"cell_type":"code","source":"centroids, cluster_assignments = kmedoids(kmer_distances_xy, 3)\nplot_kmeans(kmer_distances_xy, centroids, cluster_assignments, 3)","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"**TASK 5: silhouette analysis**<br>\nIt would be nice to have a way to objectively compare clustering performance. <br>\nWe could compare the selection of k (num clusters) to find the most optimal, or could compare different clustering algorithms to see which performs better. <br><br>\n\nSilhouette analysis is one way to do this. \nSilhouette analysis reports information about how far a datapoint is from the clusters is isn't a part of.\nIt produces a plot where we can see the separation between clusters (higher is better) \nSilhouette coefficients are produced in the range [-1, 1], where 1 indicates the the point is far from other clusters, 0 represents a point which is near the decision boundary between two clusters, and negative values indicate that the point may have been assigned to the wrong cluster. \nmore information is available at https://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_silhouette_analysis.html<br><br>\nThe cell below gives a visual representation of the silhouette scores for points in each cluster. ","metadata":{}},{"cell_type":"code","source":"from utilities import silhouette_plot\nprint('\\n\\n KMEANS')\ncentroids, cluster_assignments = kmeans(kmer_distances_xy, 2)\nplot_kmeans(kmer_distances_xy, centroids, cluster_assignments, 2)\nsilhouette_plot(kmer_distances_xy, cluster_assignments, 'kmeans')\n\nprint('\\n\\n KMEDOIDS')\ncentroids, cluster_assignments = kmedoids(kmer_distances_xy, 3)\nplot_kmeans(kmer_distances_xy, centroids, cluster_assignments, 3)\nsilhouette_plot(kmer_distances_xy, cluster_assignments, 'kmedoids')","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Let's get the silhouette scores for a run of kmeans and kmedoids. <br>\nThe average silhouette score for points in a cluster will be used as a metric to compare \nclustering performance (higher is better). <br>\nThe cell below will run k-means and k-mediods 10 times each and record average silhouette scores for each cluster. \nThis will account for a range of starting centroid locations. ","metadata":{}},{"cell_type":"code","source":"from utilities import silhouette_score\ndef silhouette_score_run(distance_matrix, k):\n    kmeans_scores = []\n    kmedoids_scores = []\n\n    while len(kmeans_scores) < 10:\n        try:\n            centroids, cluster_assignments = kmeans(distance_matrix, k)\n            kmeans_scores.append(silhouette_score(distance_matrix, cluster_assignments))\n        except ValueError:\n            pass\n\n    while len(kmedoids_scores) < 10:\n        try:\n            centroids, cluster_assignments = kmedoids(distance_matrix, k)\n            kmedoids_scores.append(silhouette_score(distance_matrix, cluster_assignments))\n        except ValueError:\n            pass\n\n    print('{:^15}{:^15}'.format('kmeans', 'kmedoids'))\n    for score1, score2 in zip(kmeans_scores, kmedoids_scores):\n        print('{:^15.2f}{:^15.2f}'.format(score1, score2))\n\n\nprint('\\nsilhouette scores: k = 2')\nsilhouette_score_run(kmer_distances_xy, 2)\nprint('\\nsilhouette scores: k = 3')\nsilhouette_score_run(kmer_distances_xy, 3)","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"In light of the silhouette analysis results produced above, we can make some objective judgement about our clustering. <br>\n- Comment on which algorithm (k-means/k-medoids) performed better for k=2.\n- Comment on what number of clusters, k=2 or k=3, produced better results. \n- Comment on why the silhouette scores are changing for each run of clustering. \n<br>\n(Maximum of 150 words)","metadata":{}},{"cell_type":"markdown","source":" -- GRADED CELL (5 marks) - complete this cell --\n \n ","metadata":{}},{"cell_type":"markdown","source":"**TASK 6: Agglomerative clustering algorithm**","metadata":{}},{"cell_type":"markdown","source":"Previously, we have used partitional clustering algorithms on our data. The other main branch, hierarchical clustering has not been yet explored. \n\nPartitional and hierarchical clustering approaches perform differently given the dataset. Your task is to write **in plan english** an agglomerative hierarchical clustering algorithm. Be specific - the exact process of merging groups and the method used to calculate distance must be mentioned, as well as what is expected as input and output. \n\nWrite your clustering algorithm in the markdown cell below.","metadata":{}},{"cell_type":"markdown","source":" -- GRADED CELL (8 marks) - complete this cell --\n \n ","metadata":{}},{"cell_type":"markdown","source":"Comment on which method, divisive or agglomerative, you think would be best given this dataset.\n\nUse information from the kmer/smith-waterman distance heatmaps, the 2D scatterplot representation of the data, and silhouette analysis. Include a comment about outliers in your answer. \n\nWrite your answer in the markdown cell below. ","metadata":{}},{"cell_type":"markdown","source":" -- GRADED CELL (5 marks) - complete this cell --\n \n ","metadata":{}}]}